Dynamic circuits updates for June (#3494)
Qiskit/documentation: docs/guides/stretch.mdx

---
title: Deferred timing resolution using stretch
description: How to use stretch to defer timing resolution with dynamic circuits
---

# Deferred timing resolution using stretch

The OpenQASM 3 language specification contains a `stretch` type that lets you specify relative timing of operations instead of absolute timing. Support for `stretch` as durations for [`Delay`](/docs/api/qiskit/2.0/circuit#qiskit.circuit.Delay) instructions was added in [Qiskit v2.0.0.](/docs/api/qiskit/release-notes/2.0#200) The concrete value of a stretch duration is resolved at compile time, after the exact duration of calibrated gates are known.  The compiler tries to minimize the stretch duration, subject to timing constraints on one or more qubits. This lets you express gate designs such as evenly spacing gates (for example, to implement a higher-order echo decoupling sequence), left-aligning a sequence of gates, or applying a gate for the duration of some sub-circuit, without knowing the exact timing.

> ## Example

Author: @jyu00 requested that these examples be converted to QASM.

```python
from qiskit.circuit import QuantumCircuit, QuantumRegister, ClassicalRegister

qubits = QuantumRegister(2)
clbits = ClassicalRegister(2)
circuit = QuantumCircuit(qubits, clbits)
(q0, q1) = qubits
(c0, c1) = clbits

# Add barriers to define the boundaries
circuit.barrier()
circuit.h(q0)
circuit.measure(q0, c0)
with circuit.if_test((c0, 1)) as else_:
    circuit.h(q0)
with else_:
    circuit.x(q0)

# Apply an XX DD sequence with stretch on qubit 1
s = circuit.add_stretch("s")
circuit.delay(s, q1)
circuit.x(q1)
circuit.delay(expr.mul(s, 2), q1)
circuit.x(q1)
circuit.delay(s, q1)
circuit.barrier()
```

Here is another example of using `stretch` to ensure a sequence of gates between two barriers are left-aligned, whatever their actual durations are:
```python
from qiskit import QuantumCircuit
from numpy import pi

qc = QuantumCircuit(5)
qc.barrier()
qc.cx(0, 1)
qc.u(pi/4, 0, pi/2, 2)
qc.cx(3, 4)

a = qc.add_stretch("a")
b = qc.add_stretch("b")
c = qc.add_stretch("c")

# Use the stretches as Delay duration.
qc.delay(a, [0, 1])
qc.delay(b, 2)
qc.delay(c, [3, 4])
qc.barrier()
```

> ## Limitations

Reviewer: Same as before, that I think this might be more readable using tabs for each of the valid/invalid set

The following constraints apply when using `stretch`:

- At most one stretch variable per _qubit set_ between barriers (implicit and explicit). A qubit set is one or more qubits; these sets must be mutually exclusive.
    Valid:

        ```python
        stretch a;
        stretch b;
        delay[a] $0, $1;
        delay[b] $2
        ```

    Invalid:

        ```python
        stretch a;
        stretch b;
        delay[a] $0, $1;
        delay[b] $0;
        ```

> - A stretch variable cannot be used between multiple sets of barriers.

Reviewer: ```suggestion
- We call the area surrounded by a set of barriers a "barrier region". A stretch variable cannot be used in multiple barrier regions.
```

        Valid:

        ```python
        barrier $0, $1;
        delay[a] $0;
        barrier $3;
        delay[a] $0;

>         barrier $0, $1;

Reviewer: ```python
# Stretch a is used inside a barrier region on q0 and q1
a = circuit.add_stretch("a")
circuit.barrier((q0, q1))
circuit.delay(a, q0)
circuit.barrier(q2)
circuit.delay(a, q0)
circuit.barrier((q0, q1))
```

I think including visualization would also help

<img width="983" height="210" alt="image" src="https://github.com/user-attachments/assets/69aa6151-5048-4e00-b131-fbd91f1f51a3" />

        ```

    Invalid:

        ```python
        barrier $0, $1;
        delay[a] $0;
        barrier $0, $1;
        delay[a] $0;
        barrier $0, $1;

>          ```

Reviewer: ```python
# Here stretch a is used in 2 barrier regions
a = circuit.add_stretch("a")
circuit.barrier((q0, q1))
circuit.delay(a, q0)
circuit.barrier((q0, q1))
circuit.delay(a, q0)
circuit.barrier((q0, q1))
```

<img width="1039" height="197" alt="image" src="https://github.com/user-attachments/assets/6c8b1177-7459-443b-b28a-f73d587e9453" />

- Stretch expressions are limited to those of the form `A*stretch + B` where `A` and `B` are floating point or integer constants.
        Valid:

        ```python

>         delay[2 * a + 3.0] $0;

Reviewer: @jakelishman do you know how to do this in Qiskit? I tried `expr.add(expr.mul(a, 2), 3.0)` but that didn't work. 

Actually I think it's the `add`, because  a simple `expr.add(a, 2)` also doesn't work.

Reviewer: ah yes this makes sense. Since I can do `delay(3)` with an implicit unit of `dt`, and a stretch also has a unit of `dt` (once resolved), I thought `expr.add(a, 2)` would work.

Reviewer 2: The Qiskit `Delay` instruction predates all the classical-expression handling, and it (and tbh all the rest of Qiskit) never really handled dimensioned quantities safely - lots of the code frequently forgot to check the unit, which was stored completely separately to the amount, even though the unit is an absolutely critical component of the quantity itself - the amount is meaningless without a unit.  The new `qiskit.circuit.Duration` represents a duration literal safely, including its literal (like OQ3 also always requires a suffix denoting the unit of duration literals).  The old handling of `Delay` will eventually be phased out.

        delay[a / 5 - 40.0] $0;
        delay[a + a] $0; # equivalent to "delay[2 * a] $0;"
        ```

    Invalid:

        ```python
        delay[2 * a * a] $0;
         ```

- Stretch expressions can only include a single stretch variable.
    Invalid:

        ```python
        stretch a;stretch b;

>         delay[ a + b ] $0;delay[ a * b ] $0;

Reviewer: ```python
a = circuit.add_stretch("a")
b = circuit.add_stretch("b")
circuit.delay(expr.add(a, b), q0)
circuit.delay(expr.mul(a, b), q0)
```

Side question also for @jakelishman, `expr.mul(a, b)` fails in Qiskit with `cannot multiply two durations`... why can't I multiple 2 durations ?

Reviewer: The intent is to do a*b in dt, not "dt squared" - e.g. use `b` as a scalar without a unit. But I see your point - it'd need to somehow cast `b` to a scalar value first.

         ```

- Stretch expressions cannot resolve to negative delay values.
    Invalid:

        ```python
        delay[a  - 40.0] $0; #if "a" resolves to a value <= 40.0
         ```

- Any remainder resulting from a stretch resolution is added to the first delay that uses the stretch.
    Example:

        ```python
        stretch a;
        barrier $0, $1;
        delay[100dt] $0;
        delay[a] $1; # resolves to 26
        x $1; # duration: 8
        delay[a] $1; # resolves to  25
        x $1; # duration: 8
        delay[a] $1; # resolves to  25
        x $1; # duration: 8

>         barrier  $0, $1;

Reviewer: ```python
a = circuit.add_stretch("a")
circuit.barrier(q0, q1)
circuit.delay(100, q0)
circuit.delay(a, q1)  # resolve to 26
circuit.x(q1)  # duration: 8
circuit.delay(a, q1)  # resolve to 25
circuit.x(q1)  # duration: 8
circuit.delay(a, q1)  # resolve to 25
circuit.x(q1)  # duration: 8
circuit.barrier(q0, q1)
```

```python
from qiskit import QuantumCircuit
from numpy import pi

qc = QuantumCircuit(5)
qc.barrier()
qc.cx(0, 1)
qc.u(pi/4, 0, pi/2, 2)
qc.cx(3, 4)

a = qc.add_stretch("a")
b = qc.add_stretch("b")
c = qc.add_stretch("c")

# Use the stretches as Delay duration.
qc.delay(a, [0, 1])
qc.delay(b, 2)
qc.delay(c, [3, 4])
qc.barrier()
```

<Admonition type="note">
 Any remainder resulting from a stretch resolution is added to the first delay that uses the stretch.

    Example:

        ```python
        a = circuit.add_stretch("a")
        circuit.barrier(q0, q1)
        circuit.delay(100, q0)
        circuit.delay(a, q1)  # resolve to 26
        circuit.x(q1)  # duration: 8
        circuit.delay(a, q1)  # resolve to 25
        circuit.x(q1)  # duration: 8
        circuit.delay(a, q1)  # resolve to 25
        circuit.x(q1)  # duration: 8
        circuit.barrier(q0, q1)
         ```

    The above code resolves to a value of 25 with a remainder of 1. The first delay`[a]` will have the remainder added.

    Stretch resolution equation:  $a + 8 + a + 8 + a + 8 = 100   =   3*a + 24$
</Admonition>

> ## Limitations

Reviewer: ```suggestion
## Qiskit Runtime Limitations
```

        circuit.barrier((q0, q1))
        ```

        ![Illustration of the previous code output](/docs/images/guides/stretch/good-barrier-region.png 'Valid use of stretch in a barrier region')

     </TabItem>
    </Tabs>

- Stretch expressions are limited to those of the form `A*stretch + B` where `A` and `B` are floating point or integer constants.
    <Tabs>
         <TabItem value="Invalid" label="Invalid">
        ```python
        delay[2 * a * a] $0;
         ```
     </TabItem>
     <TabItem value="Valid" label="Valid">
        ```python
        qubit[1] q;
        stretch a;
        delay[a * 2 + 3dt] q[0];
        ```
     </TabItem>
    </Tabs>

- Stretch expressions can only include a single stretch variable.
    <Tabs>
        <TabItem value="Invalid" label="Invalid">
        ```python
        qubit[1] q;
        stretch a;
        stretch b;
        delay[a * (b / 1dt)] q[0];
         ```
     </TabItem>
         <TabItem value="Valid" label="Valid">
        ```python
        xxxxxx
        ```
     </TabItem>

    </Tabs>

- Stretch expressions cannot resolve to negative delay values.
    <Tabs>
         <TabItem value="Invalid" label="Invalid">
        ```python

>         delay[a  - 40.0] $0; #if "a" resolves to a value <= 40.0

Reviewer: ```suggestion
        delay[a  - 40.0] $0; #if "a" resolves to a value <= 40.0
        from qiskit.circuit import Duration

        circuit.barrier((q0, q1))
        circuit.delay(20, q1)
        # The length of this barrier region is 20dt, meaning the
        # equation for solving stretch 'a' is a + 40dt = 20dt, giving a = -20dt. 
        circuit.delay(expr.add(a, Duration.dt(40)), q0)
        circuit.barrier((q0, q1))        
```

         ```
     </TabItem>
     <TabItem value="Valid" label="Valid">
        ```python

>         xxxxxx

Reviewer: ```suggestion
        circuit.barrier((q0, q1))
        circuit.delay(20, q1)
        circuit.delay(a, q0)
        circuit.barrier((q0, q1))             
```

        ```
     </TabItem>

>     </Tabs>

Reviewer 2: 1. `a - 40.0` should probably be `a - 40dt` - you can't add "just a number" to a quantity (`a`) with dimension "time".
2. I'm not clear what this restriction actually is - in the OpenQASM 3 spec, a duration of `a - 40dt` being used is valid and implies that `a >= 40dt`, and the stretch resolver must account for that.  It wouldn't be a solution if `a` resolved to a value `< 40dt` in this situation (note: the comment says `<=` but it probably means `<`; `0dt` presumably _is_ a valid duration).

Reviewer 3: The stretch resolution logic is currently not sophisticated enough to do that. So, if you write:
```
barrier;
stretch a;
delay[a - 40dt] $0;
delay[20dt] $1;
barrier;
```
The solver will not "inflate" the region between the barriers to ensure that `a` is non-negative. Instead, it will just solve `a - 40dt == 20dt` giving `a = -20dt`.

Author: If you solve `a - 40dt == 20dt`, wouldn't you get  `a = 60dt` ?  Or does something special happen that I don't understand?

Reviewer 2: Ok ace, thanks - I think an explanation based on your example would make things clearer.

Reviewer 2: Becky, yeah, you're right - I think the example would be `a + 40dt` and `20dt`.

- Stretch expressions are limited to those of the form `X*stretch + Y` where `X` and `Y` are floating point or integer constants.
    <Tabs>
         <TabItem value="Invalid" label="Invalid">
        ```python
        a = circuit.add_stretch("a")
        b = circuit.add_stretch("b")
        c = circuit.add_stretch("c")

        # (a /  b) * c is not supported
        circuit.delay(expr.mul(expr.div(a, b), c), q1)        
         ```
     </TabItem>
     <TabItem value="Valid" label="Valid">
        ```python
        from qiskit.circuit import Duration
       
        a = circuit.add_stretch("a")
        circuit.delay(expr.add(expr.mul(a, 2), Duration.dt(3)), 0)        
        ```
     </TabItem>
    </Tabs>

- Stretch expressions can only include a single stretch variable.
    <Tabs>
        <TabItem value="Invalid" label="Invalid">
        ```python
        a = circuit.add_stretch("a")
        b = circuit.add_stretch("b")
        circuit.delay(expr.add(a, b), 0)
         ```
     </TabItem>
         <TabItem value="Valid" label="Valid">
        ```python
        a = circuit.add_stretch("a")
        circuit.delay(expr.add(a, a), 0)
        ```
     </TabItem>

    </Tabs>

> - Stretch expressions cannot resolve to negative delay values.

Reviewer 3: Maybe just add an extra comment that the current solver doesn't infer non-negativity constraints?

---
title: Deferred timing resolution using stretch
description: How to use stretch to defer timing resolution with dynamic circuits
---

# Deferred timing resolution using stretch

> The OpenQASM 3 language specification contains a `stretch` type that lets you specify relative timing of operations instead of absolute timing. Support for `stretch` as durations for [`Delay`](/docs/api/qiskit/circuit#qiskit.circuit.Delay) instructions was added in [Qiskit v2.0.0.](/docs/api/qiskit/release-notes/2.0#200) The concrete value of a stretch duration is resolved at compile time, after the exact duration of calibrated gates are known.  The compiler tries to minimize the stretch duration, subject to timing constraints on one or more qubits. This lets you express gate designs such as evenly spacing gates (for example, to implement a higher-order echo decoupling sequence), left-aligning a sequence of gates, or applying a gate for the duration of some sub-circuit, without knowing the exact timing.

Reviewer 4: ```suggestion
The OpenQASM 3 language specification contains a `stretch` type with which you can specify relative timing of operations instead of absolute timing. Support for `stretch` as durations for [`Delay`](/docs/api/qiskit/circuit#qiskit.circuit.Delay) instructions was added in [Qiskit v2.0.0.](/docs/api/qiskit/release-notes/2.0#200) The concrete value of a stretch duration is resolved at compile time, after the exact duration of calibrated gates are known.  The compiler tries to minimize the stretch duration, subject to timing constraints on one or more qubits. You can then express gate designs such as evenly spacing gates (for example, to implement a higher-order echo decoupling sequence), left-aligning a sequence of gates, or applying a gate for the duration of some sub-circuit, without knowing the exact timing.
```

```python
from qiskit import QuantumCircuit
from numpy import pi

qc = QuantumCircuit(5)
qc.barrier()
qc.cx(0, 1)
qc.u(pi/4, 0, pi/2, 2)
qc.cx(3, 4)

a = qc.add_stretch("a")
b = qc.add_stretch("b")
c = qc.add_stretch("c")

# Use the stretches as Delay duration.
qc.delay(a, [0, 1])
qc.delay(b, 2)
qc.delay(c, [3, 4])
qc.barrier()
```

<Admonition type="note">
 When using `stretch` with Qiskit Runtime, any remainder resulting from a stretch resolution is added to the first delay that uses the stretch.

    Example:

        ```python
        a = circuit.add_stretch("a")
        circuit.barrier(q0, q1)
        circuit.delay(100, q0)
        circuit.delay(a, q1)  # resolve to 26
        circuit.x(q1)  # duration: 8
        circuit.delay(a, q1)  # resolve to 25
        circuit.x(q1)  # duration: 8
        circuit.delay(a, q1)  # resolve to 25
        circuit.x(q1)  # duration: 8
        circuit.barrier(q0, q1)
         ```

    The above code resolves to a value of 25 with a remainder of 1. The first delay`[a]` will have the remainder added.

    Stretch resolution equation:  $a + 8 + a + 8 + a + 8 = 100   =   3*a + 24$
</Admonition>

> ## Qiskit Runtime Limitations

Reviewer 4: ```suggestion
## Qiskit Runtime limitations
```
